File: dart/lib/src/bindings/delete_name_registry.dart
import 'package:solana/solana.dart' hide RpcClient;

import '../constants/addresses.dart';
import '../instructions/delete_name_registry_instruction.dart';
import '../instructions/instruction_types.dart';
import '../rpc/rpc_client.dart';
import '../states/registry.dart';
import '../utils/get_domain_key_sync.dart';

/// Parameters for deleting a name registry
class DeleteNameRegistryParams {
  const DeleteNameRegistryParams({
    required this.name,
    required this.refundTargetKey,
    this.nameClass,
    this.nameParent,
  });

  /// The name of the name account
  final String name;

  /// The refund destination address
  final Ed25519HDPublicKey refundTargetKey;

  /// The class of this name, if it exists
  final Ed25519HDPublicKey? nameClass;

  /// The parent name of this name, if it exists
  final Ed25519HDPublicKey? nameParent;
}

/// Delete the name account and transfer the rent to the target.
///
/// This function mirrors js/src/bindings/deleteNameRegistry.ts
///
/// [rpc] - The solana connection object to the RPC node
/// [params] - The parameters for deleting the name registry
///
/// Returns the transaction instruction for deleting the name registry
Future<TransactionInstruction> deleteNameRegistry(
  RpcClient rpc,
  DeleteNameRegistryParams params,
) async {
  final result = getDomainKeySync(params.name);

  Ed25519HDPublicKey nameOwner;
  if (params.nameClass != null) {
    nameOwner = params.nameClass!;
  } else {
    final domainKeyResult = await result;
    final state = await RegistryState.retrieve(rpc, domainKeyResult.pubkey);
    nameOwner = Ed25519HDPublicKey.fromBase58(state.owner);
  }

  final domainKeyResult = await result;
  final instruction = DeleteNameRegistryInstruction()
    ..setParams(
      programAddress: nameProgramAddress,
      domainAddress: domainKeyResult.pubkey,
      refundTarget: params.refundTargetKey.toBase58(),
      owner: nameOwner.toBase58(),
    );

  return instruction.build();
}

File: js/src/bindings/deleteNameRegistry.ts
import { Connection, PublicKey, TransactionInstruction } from "@solana/web3.js";
import { deleteInstruction } from "../instructions/deleteInstruction";
import { NameRegistryState } from "../state";
import { NAME_PROGRAM_ID } from "../constants";
import { getHashedNameSync } from "../utils/getHashedNameSync";
import { getNameAccountKeySync } from "../utils/getNameAccountKeySync";

/**
 * Builds an instruction to delete a name account and transfer reclaimed rent.
 *
 * @param connection Solana RPC connection
 * @param name Name of the name account
 * @param refundTargetKey Refund destination address
 * @param nameClass Optional class of the name account
 * @param nameParent Optional parent name account
 * @returns Transaction instruction.
 */
export async function deleteNameRegistry(
  connection: Connection,
  name: string,
  refundTargetKey: PublicKey,
  nameClass?: PublicKey,
  nameParent?: PublicKey,
): Promise<TransactionInstruction> {
  const hashed_name = getHashedNameSync(name);
  const nameAccountKey = getNameAccountKeySync(
    hashed_name,
    nameClass,
    nameParent,
  );

  let nameOwner: PublicKey;
  if (nameClass) {
    nameOwner = nameClass;
  } else {
    nameOwner = (await NameRegistryState.retrieve(connection, nameAccountKey))
      .registry.owner;
  }

  const changeAuthoritiesInstr = deleteInstruction(
    NAME_PROGRAM_ID,
    nameAccountKey,
    refundTargetKey,
    nameOwner,
  );

  return changeAuthoritiesInstr;
}
